### ISSUE TITLE
[FIXED] HTTP 400: Bad Request when asking for Refresh Token and Access Token

### ISSUE THREAD

**Issue by ShaminAsfaq** [spotify/web-api#1410]
**[UPDATE]**: 

> FIXED !

I had to do two things:

1. Had a typo which was `Autorizacion`, fixed it.
2. The `POST` request in axios requires the body to be a string. So, I had to `stringify` it. Tried `JSON.stringify()`, but it didn't work as it was adding some backslashes (\\). Issue resolved when I used `querystring`.

```
npm i querystring
------------------------

const qs = require('querystring');

//and in axios,

data: qs.stringify({
         grant_type: `${grant_type}`,
         code: `${authorizationCode}`,
         redirect_uri: `${redirectUri}`
}),

```

Issue found on February 19th 2015.

Hello,

This is my first issue creation on GitHub ever. Please consider this in case I make any mistakes.

I am trying to generate Refresh Token using the `spotify-web-api-node`. Here's what I have tried:

```
const URL = 'https://accounts.spotify.com/api/token';
const clientId = 'e049336609c1483e93b27a63bdefa50b';
const clientSecret = '78e9cfed93fa4cfeacfa1bb8dcd9a821';
const grant_type = 'authorization_code';
const redirectUri = 'http://118.179.95.206:5000';
const base64 = Buffer.from(clientId + ':' + clientSecret).toString('base64');

// Below code is what I got from the URL after I was redirected to the URL
// authorizationCode = 'AQB07r224qlj1N66Ytm7_Ly-SA2pRF3CLktf3iIMK2zXrplPERdsFMzEBtA58n8dElolGFWB6p0Ui1SPGaerbfYMapdudO7j1mHZ8KGjuUvH5sNbdU_RvCvFYCDjrB9WF3AekWwzoVAYC4rypf3f0KoqFBjRCoHaa6XU7JDQjVuVlPwymRzNBtXNZfgzXNRU5SO8OQVxcVmgAi715RCSamuOtLU4sCDgJyMDQx4';



// First I tried this line below:
// var response = await axios.post(URL, { data: { grant_type: `${grant_type}`, code: `${authorizationCode}`, redirect_uri: `${redirectUri}` },  headers: { "Content-type": "application/x-www-form-urlencoded", "Authorization": 'Basic ' + base64 } })

// Then tried this line:
var response = axios({ method: 'POST', url: URL,
                headers: {
                    'Autorizacion': `Basic ${base64}`,
                    'Content-Type': 'application/x-www-form-urlencoded'
                }, 
                data: {
                     grant_type: `${grant_type}`,
                     code: `${authorizationCode}`,
                     redirect_uri: `${redirectUri}`
                } 
}).then().catch((err) => {
    res.send(err)
})

```

Here's the Image to the settings of the app I created from Spotify Dashboard: [Settings](https://ibb.co/vxYPy3P)

#### Endpoint(s):
* `POST https://accounts.spotify.com/api/token`

#### Scope(s):
* None (I didn't understand this)

#### Steps to reproduce:
1. Please follow the exact codes that I wrote. I hope you will face the same problem.

#### Expected behaviour:
Access Token and Refresh Token

#### Actual behaviour:
400 Bad Request

Feel free to instruct me if this issue is missing some information that should have been here.

Thanks!

**Comment by 0xNF** [comment]
First things first, go and regenerate your `Client Secret`. You shouldn't post that particular string to the public, because bad actors can now impersonate your application.

Second, it looks like you localized the string `Autorizacion`. It should be `Authorization`.

There may be more errors, I just gave it a quick lookthrough, but try fixing that authorization spelling thing and see what happens.

**Comment by ShaminAsfaq** [comment]
> First things first, go and regenerate your `Client Secret`. You shouldn't post that particular string to the public, because bad actors can now impersonate your application.
> 
> Second, it looks like you localized the string `Autorizacion`. It should be `Authorization`.
> 
> There may be more errors, I just gave it a quick lookthrough, but try fixing that authorization spelling thing and see what happens.

Hello 0xNF!

Thanks for the reply. I re-generated the client secret and fixed the typo. The problem still seem to exist. I get this from the catch block:

```
{
"message": "Request failed with status code 400",
"name": "Error",
"stack": "Error: Request failed with status code 400\n    at createError (/home/shamin/Git/react-fiesta/shamin-lives/server/node_modules/axios/lib/core/createError.js:16:15)\n    at settle (/home/shamin/Git/react-fiesta/shamin-lives/server/node_modules/axios/lib/core/settle.js:17:12)\n    at IncomingMessage.handleStreamEnd (/home/shamin/Git/react-fiesta/shamin-lives/server/node_modules/axios/lib/adapters/http.js:237:11)\n    at IncomingMessage.emit (events.js:203:15)\n    at endReadableNT (_stream_readable.js:1129:12)\n    at process._tickCallback (internal/process/next_tick.js:63:19)",
"config": {
"url": "https://accounts.spotify.com/api/token",
"method": "post",
"data": "{\"grant_type\":\"authorization_code\",\"code\":\"AQBpzxvKrcSAgFhpqfWMzbUk3acZNrAHc_jDm3QU8vSpVjL1IEgnKHkdapfu8v2N9DOlD52RWC6ICRVvSHucDHimNv6XVaOk_YEWuYgYOzLrDCLGLtZnuGZ6zEG08-PaUIUW8SLAUIhtladP66xpYGiffT_XnERNSoBXyWZR0ivBkA9_HJGlpkBopUVr4-Wr18twYDJ8N00aXnh-fbW9i_dQY_2BAcrE8zsf4yA\",\"redirect_uri\":\"http://118.179.95.206:5000\"}",
"headers": {
"Accept": "application/json, text/plain, */*",
"Content-Type": "application/x-www-form-urlencoded",
"Authorization": "Basic ZTA0OTMzNjYwOWMxNDgzZTkzYjI3YTYzYmRlZmE1MGI6ZDQwNGE3MWRkYzc3NDc4YWI4OWE2NWI0YTliYWE4OGE=",
"User-Agent": "axios/0.19.0",
"Content-Length": 320
},
"transformRequest": [
null
],
"transformResponse": [
null
],
"timeout": 0,
"xsrfCookieName": "XSRF-TOKEN",
"xsrfHeaderName": "X-XSRF-TOKEN",
"maxContentLength": -1
}
}
```

and this from my terminal:

```
(node:27761) UnhandledPromiseRejectionWarning: Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:470:11)
    at ServerResponse.header (/home/shamin/Git/react-fiesta/shamin-lives/server/node_modules/express/lib/response.js:771:10)
    at ServerResponse.send (/home/shamin/Git/react-fiesta/shamin-lives/server/node_modules/express/lib/response.js:170:12)
    at ServerResponse.json (/home/shamin/Git/react-fiesta/shamin-lives/server/node_modules/express/lib/response.js:267:15)
    at ServerResponse.send (/home/shamin/Git/react-fiesta/shamin-lives/server/node_modules/express/lib/response.js:158:21)
    at app.get (/home/shamin/Git/react-fiesta/shamin-lives/server/script.js:124:9)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:27761) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:27761) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

```

**Comment by 0xNF** [comment]
Glad you solved it. Would you consider posting the solution? In case others stumble upon this page and have the same error

**Comment by ShaminAsfaq** [comment]
> Glad you solved it. Would you consider posting the solution? In case others stumble upon this page and have the same error

Surely!

Although I have edited the main issue description, I am commenting it here, again.

**_My first problem_** was a `typo`. I had to pass in a header for **_AUTHORIZATION_** and I made a silly typing mistake which was: `Autorizacion`. So, keep eye on your spellings.

**_My second problem was_**, I did not know that I have to pass in the data as a string in the `POST` method. I was sending in an object like this:

```
data: {
    grant_type: `${grant_type}`,
    code: `${authorizationCode}`,
    redirect_uri: `${redirectUri}`
} 
```
wherein the actual `data` should be one single string. To make that happen, I first tried `JSON.stringify(myObject)`, but I failed to make it work. Because `JSON.stringify()` adds backslashes (`\`) in many places of the object I was trying to pass in the `POST` request, and it resulted in `HTTP 400: Bad Request` for me.

The solution I found involved `querystring`. The sysntax to `stringify` the object is similar to the previous one I just wrote 2 lines above. Here's how to do it: 

> **Installation**: npm i querystring
-----------------------------------------------

> **Inside code**:
-------------------

```
const qs = require('querystring');

//and in axios,
data: qs.stringify({
         grant_type: `${grant_type}`,
         code: `${authorizationCode}`,
         redirect_uri: `${redirectUri}`
}),
```

Look carefully how I put the object inside the `qs.stringify()` method. Previously, it was just:

```
data: {
    grant_type: `${grant_type}`,
    code: `${authorizationCode}`,
    redirect_uri: `${redirectUri}`
},
```
but now it resides inside the `qs.stringify()` method.

That was it for me. 

> Problem Solved!
